lib/imap/providers/gmail.rb

```diff
@@ -0,0 +1,148 @@
+# frozen_string_literal: true
+
+module Imap
+  module Providers
+    class Gmail < Generic
+      X_GM_LABELS = 'X-GM-LABELS'
+
+      def imap
+        @imap ||= begin
```

Comment:
Might be able to simplify this: `@imap ||= super.tap { |imap| apply_gmail_patch(imap) }`